x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/action/RollupShardTask.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.rollup.action;
+
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.tasks.CancellableTask;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.xpack.core.downsample.DownsampleConfig;
+import org.elasticsearch.xpack.core.rollup.RollupField;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class RollupShardTask extends CancellableTask {
+    private String rollupIndex;
+    private DownsampleConfig config;
+    private final RollupShardStatus status;
+    private final AtomicLong numReceived = new AtomicLong(0);
+    private final AtomicLong numSent = new AtomicLong(0);
+    private final AtomicLong numIndexed = new AtomicLong(0);
+    private final AtomicLong numFailed = new AtomicLong(0);
+
+    public RollupShardTask(
+        long id,
+        String type,
+        String action,
+        TaskId parentTask,
+        String rollupIndex,
+        DownsampleConfig config,
+        Map<String, String> headers,
+        ShardId shardId
+    ) {
+        super(id, type, action, RollupField.NAME + "_" + rollupIndex + "[" + shardId.id() + "]", parentTask, headers);
+        this.rollupIndex = rollupIndex;
+        this.config = config;
+        this.status = new RollupShardStatus(shardId, numReceived, numSent, numIndexed, numFailed);
+    }
+
+    public String getRollupIndex() {
+        return rollupIndex;
+    }
+
+    public DownsampleConfig config() {
+        return config;
+    }
+
+    @Override
+    public Status getStatus() {
+        return status;
```

Comment:
Instead of creating a status object in the constructor. The status object should be created in this method.
This way a read only copy of the status of the task is created. And then there is no need to pass the atomic longs to the status.

Reply:
Yeah, I change the code.

Reply:
This method should return a new status instance:

```
new RollupShardStatus(shardId, numReceived, numSent, numIndexed, numFailed);
```

and then the status field in this class can be removed.